tutorials/preprocessing/35_artifact_correction_regression.py

```diff
@@ -23,146 +24,195 @@
-# trigger times on detected sound onsets on channel UADC001-4408.
-
-events = mne.find_events(raw, stim_channel='UPPT001')
-sound_data = raw[raw.ch_names.index('UADC001-4408')][0][0]
-onsets = np.where(np.abs(sound_data) > 2. * np.std(sound_data))[0]
-min_diff = int(0.5 * raw.info['sfreq'])
-diffs = np.concatenate([[min_diff + 1], np.diff(onsets)])
-onsets = onsets[diffs > min_diff]
-assert len(onsets) == len(events)
-events[:, 0] = onsets
-epochs = mne.Epochs(raw, events, event_id=1, decim=decim, preload=True)
+# We can see there is some EOG activity that is likely bleeding into the EEG
+# evoked response. At around 250ms this becomes especially noticeable. Let's
+# apply regression to subtract the EOG signal from the EEG signals to clean it
+# up.
 
 # %%
 # Compute and apply EOG regression
 # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-# Next we'll compare the `~mne.Evoked` data (average across epochs) before and
-# after we regress out the EOG signal. We do this by first estimating the
-# regression coefficients on epoch data with the evoked response subtracted
-# out. As long as the EOG artifacts are in the epochs, then the algorithm
-# should be able to estimate regression coefficients. The EOG artifacts do not
-# need to be time-locked to the stimulus/epoch event timing of the epochs.
+# Now, we'll compare the evoked response before and after we regress out the
+# EOG signal. First, let's try plain regression, and then we'll explore more
+# advanced techniques.
 
-# do regression
-_, betas = mne.preprocessing.regress_artifact(epochs.copy().subtract_evoked())
+# Perform regression using the EOG sensor as independent variable and the EEG
+# sensors as dependent variables.
+model_plain = EOGRegression(picks='eeg', picks_artifact='eog').fit(epochs)
+model_plain.plot(vmax=0.4)  # Show the regression coefficients as a topomap
 
 # %%
-# We then use those coefficients to remove the EOG signal from the original
-# data:
-
-epochs_clean, _ = mne.preprocessing.regress_artifact(epochs, betas=betas)
+# The regression coefficients shows the linear relationship between each EEG
+# sensor and the EOG sensor. Note that occipital sensors have a positive
+# relationship, as we set a common-average reference when we loaded the data
+# above.
+#
+# Now we are ready to use these coefficients to subtract the EOG signal from
+# the EEG signals.
+epochs_clean_plain = model_plain.apply(epochs)
+fig = epochs_clean_plain.average('all').plot(**plot_kwargs)
+fig.set_size_inches(6, 6)
 
 # %%
-# Visualize the effect on auditory epochs
-# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-# Now we can plot the auditory evoked response before and after regression:
-
-# get ready to plot
-plot_picks = ['meg', 'eog']
-evo_kwargs = dict(picks=plot_picks, spatial_colors=True,
-                  verbose='error')  # ignore warnings about spatial colors
+# Regressing the EOG signal out of the EEG signals has reduced the peak around
+# 250ms that was partly there because of eye artifacts.
```

Comment:
This is extremely miniscule and hard to see – maybe you should mention this in the text? Otherwise, people might go looking for the difference and wonder why they can't see it. Maybe you could even add a red arrow pointing to the difference, that would be great.

Also, same as I mentioned before, there should be a space between 250 and ms (I won't comment on this if this comes up again).

Reply:
I now realize we should re-apply the baseline after regression. Now the effect is much more noticeable!

Reply:
Had to drop the spacial colors due to warnings. Now it's again less noticeable, but still noticeable IMO, WDYT?

Reply:
I think it's OK, if you look a bit closer then it's clearly visible!